File: 09265.cpp
#include <cstdio>
#define MAX 10000

unsigned parseInt(){
	unsigned ret = 0;
	for(int bit = 0; bit < 32; ++bit){
		char c = getchar();
		ret += (c - '0') * (1U << bit);
	}
 	return ret;
}

int main(){
	unsigned n = parseInt(), m = parseInt();
	unsigned a[MAX][MAX];
	for(unsigned i = 0; i < n; ++i){
		for(int j = 0; j < m; ++j){
			a[i][j] = parseInt();
		}
	}
	unsigned low, high;
	while(cin >> low >> high){
		int ans = 0;
		int left = m - 1, right = m - 1;
		for(int i = 0; i < n; ++i){
			while(left >= 0 && a[i][left] >= low){
				--left;
			}
			while(right >= 0 && a[i][right] > high){
				--right;
			}
			ans += right - left;
		}
		cout << ans << '\n';
	}
	return 0;
}

File: 09265.py
import sys


def read_int():
    data = sys.stdin.buffer.read(4)
    # print(data)
    if data:
        return int.from_bytes(data, "little")
    else:
        raise EOFError


def main():
    n = read_int()
    m = read_int()
    a = []
    for i in range(n):
        a.append([0] * m)
        for j in range(m):
            a[i][j] = read_int()
    while True:
        try:
            low = read_int()
            high = read_int()
        except EOFError:
            break
        ans = 0
        left, right = len(a[0]) - 1, len(a[0]) - 1
        for i in range(n):
            while left >= 0 and a[i][left] >= low:
                left -= 1
            while right >= 0 and a[i][right] > high:
                right -= 1
            ans += right - left
        sys.stdout.write("{}\n".format(ans))


main()
